pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go

```diff
@@ -0,0 +1,145 @@
+	// constraint is being dropped with a "constraint does not exist" error.
+	constraintID := constraintNameElem.ConstraintID
+	if skip, err := shouldSkipValidatingConstraint(b, tbl.TableID, constraintID); err != nil {
+		panic(err)
+	} else if skip {
+		return
+	}
+
+	// 3. Drop the not-valid constraint and old constraint name element
+	//    Add a new sibling constraint and a new constraint name element.
+	validateConstraint(b, tbl.TableID, validateConstraintSpec{
+		constraintNameElem: constraintNameElem,
+		ckNotValidElem:     retrieveCheckConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
+		uwiNotValidElem:    retrieveUniqueWithoutIndexConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
+		fkNotValidElem:     retrieveForeignKeyConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
+	})
+}
+
+func retrieveCheckConstraintUnvalidatedElem(
+	b BuildCtx, tableID catid.DescID, constraintID catid.ConstraintID,
+) (CheckConstraintUnvalidatedElem *scpb.CheckConstraintUnvalidated) {
+	scpb.ForEachCheckConstraintUnvalidated(b.QueryByID(tableID), func(
+		current scpb.Status, target scpb.TargetStatus, e *scpb.CheckConstraintUnvalidated,
+	) {
+		if e.ConstraintID == constraintID {
+			CheckConstraintUnvalidatedElem = e
+		}
+	})
+	return CheckConstraintUnvalidatedElem
+}
+
+func retrieveUniqueWithoutIndexConstraintUnvalidatedElem(
+	b BuildCtx, tableID catid.DescID, constraintID catid.ConstraintID,
+) (UniqueWithoutIndexConstraintUnvalidatedElem *scpb.UniqueWithoutIndexConstraintUnvalidated) {
+	scpb.ForEachUniqueWithoutIndexConstraintUnvalidated(b.QueryByID(tableID), func(
+		current scpb.Status, target scpb.TargetStatus, e *scpb.UniqueWithoutIndexConstraintUnvalidated,
+	) {
+		if e.ConstraintID == constraintID {
+			UniqueWithoutIndexConstraintUnvalidatedElem = e
+		}
+	})
+	return UniqueWithoutIndexConstraintUnvalidatedElem
+}
+
+func retrieveForeignKeyConstraintUnvalidatedElem(
+	b BuildCtx, tableID catid.DescID, constraintID catid.ConstraintID,
+) (ForeignKeyConstraintUnvalidatedElem *scpb.ForeignKeyConstraintUnvalidated) {
+	scpb.ForEachForeignKeyConstraintUnvalidated(b.QueryByID(tableID), func(
+		current scpb.Status, target scpb.TargetStatus, e *scpb.ForeignKeyConstraintUnvalidated,
+	) {
+		if e.ConstraintID == constraintID {
+			ForeignKeyConstraintUnvalidatedElem = e
+		}
+	})
+	return ForeignKeyConstraintUnvalidatedElem
+}
+
+type validateConstraintSpec struct {
+	constraintNameElem *scpb.ConstraintWithoutIndexName
+	ckNotValidElem     *scpb.CheckConstraintUnvalidated
+	uwiNotValidElem    *scpb.UniqueWithoutIndexConstraintUnvalidated
+	fkNotValidElem     *scpb.ForeignKeyConstraintUnvalidated
+}
+
+func validateConstraint(b BuildCtx, tableID catid.DescID, spec validateConstraintSpec) {
+	nextConstraintID := b.NextTableConstraintID(tableID)
+	if spec.ckNotValidElem != nil {
+		b.Drop(spec.ckNotValidElem)
+		b.Add(&scpb.CheckConstraint{
+			TableID:               tableID,
+			ConstraintID:          nextConstraintID,
```

Comment:
interesting solution here, so that means the constraint ID changes... I will need to handle the back-references when a UDF is used in a check constraint heh. Is it possible to use the same constraint ID here?

Reply:
oh, actually nvm. it's handled :)